From 38ccf257f278e6b71f73dee9ccb568fe2ad3037a Mon Sep 17 00:00:00 2001
From: Michal Domonkos <mdomonko@redhat.com>
Date: Fri, 31 Oct 2025 11:21:28 +0100
Subject: [PATCH 1/2] Obtain nvr string only once, before the loop

Refactor loadKeyringFromDB() a bit so that the nvr string can be reused
in multiple places. This will be handy in the next commit.

No functional change.
---
 lib/rpmts.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/lib/rpmts.c b/lib/rpmts.c
index 69888acfc..ceb021789 100644
--- a/lib/rpmts.c
+++ b/lib/rpmts.c
@@ -327,10 +327,12 @@ static int loadKeyringFromDB(rpmts ts)
     while ((h = rpmdbNextIterator(mi)) != NULL) {
 	struct rpmtd_s pubkeys;
 	const char *key;
+	char *nvr;
 
 	if (!headerGet(h, RPMTAG_PUBKEYS, &pubkeys, HEADERGET_MINMEM))
 	   continue;
 
+	nvr = headerGetAsString(h, RPMTAG_NVR);
 	while ((key = rpmtdNextString(&pubkeys))) {
 	    uint8_t *pkt;
 	    size_t pktlen;
@@ -341,9 +343,7 @@ static int loadKeyringFromDB(rpmts ts)
 		rpmPubkey *subkeys = rpmGetSubkeys(key, &subkeysCount);
 
 		if (rpmKeyringAddKey(ts->keyring, key) == 0) {
-		    char *nvr = headerGetAsString(h, RPMTAG_NVR);
 		    rpmlog(RPMLOG_DEBUG, "added key %s to keyring\n", nvr);
-		    free(nvr);
 		    nkeys++;
 		}
 		rpmPubkeyFree(key);
@@ -352,12 +352,10 @@ static int loadKeyringFromDB(rpmts ts)
 		    rpmPubkey subkey = subkeys[i];
 
 		    if (rpmKeyringAddKey(ts->keyring, subkey) == 0) {
-			char *nvr = headerGetAsString(h, RPMTAG_NVR);
 			rpmlog(RPMLOG_DEBUG,
 			    "added subkey %d of main key %s to keyring\n",
 			    i, nvr);
 
-			free(nvr);
 			nkeys++;
 		    }
 		    rpmPubkeyFree(subkey);
@@ -367,6 +365,7 @@ static int loadKeyringFromDB(rpmts ts)
 	    }
 	}
 	rpmtdFreeData(&pubkeys);
+	free(nvr);
     }
     rpmdbFreeIterator(mi);
 
-- 
2.51.1


From c68a557146c594458fb00109863565651cbe8f0a Mon Sep 17 00:00:00 2001
From: Michal Domonkos <mdomonko@redhat.com>
Date: Fri, 31 Oct 2025 14:17:25 +0100
Subject: [PATCH 2/2] Improve error handling on keystore load

Report a failure to load a key, but only as a non-fatal warning so that
other keys can still be loaded, if any. This behavior is in the spirit
of #3996, too.

Only emit *one* warning in the keystore_rpmdb, to keep the (code) noise
down. This covers all three steps (rpmBase64Decode, rpmPubkeyNew and
rpmKeyringAddKey). We don't need to check if key is NULL since that's
handled by rpmKeyringAddKey() which returns immediately if that's the
case.

In the other two backends, we already handle a failure to *read* a key,
so keep that and just add another warning for a failed addition.

This would be difficult to test (especially in the rpmdb backend) so
skimping on that here.

(backported from commit 75de4f3eb28c686e7f73b750ca116a8e57b7a690)
---
 lib/rpmts.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/lib/rpmts.c b/lib/rpmts.c
index ceb021789..f6fe38c7a 100644
--- a/lib/rpmts.c
+++ b/lib/rpmts.c
@@ -281,6 +281,7 @@ static int loadKeyringFromFiles(rpmts ts)
     }
 
     for (char **f = files; *f; f++) {
+	int rc = 1;
 	int subkeysCount, i;
 	rpmPubkey *subkeys;
 	rpmPubkey key = rpmPubkeyRead(*f);
@@ -289,7 +290,7 @@ static int loadKeyringFromFiles(rpmts ts)
 	    rpmlog(RPMLOG_ERR, _("%s: reading of public key failed.\n"), *f);
 	    continue;
 	}
-	if (rpmKeyringAddKey(ts->keyring, key) == 0) {
+	if ((rc = rpmKeyringAddKey(ts->keyring, key)) == 0) {
 	    nkeys++;
 	    rpmlog(RPMLOG_DEBUG, "added key %s to keyring\n", *f);
 	}
@@ -305,10 +306,15 @@ static int loadKeyringFromFiles(rpmts ts)
 		    i, *f);
 
 		nkeys++;
+	    } else {
+		rc = 1;
 	    }
 	    rpmPubkeyFree(subkey);
 	}
 	free(subkeys);
+
+	if (rc)
+	    rpmlog(RPMLOG_WARNING, _("Could not load key %s\n"), *f);
     }
 exit:
     free(pkpath);
@@ -334,15 +340,16 @@ static int loadKeyringFromDB(rpmts ts)
 
 	nvr = headerGetAsString(h, RPMTAG_NVR);
 	while ((key = rpmtdNextString(&pubkeys))) {
+	    int rc = 1;
 	    uint8_t *pkt;
 	    size_t pktlen;
 
-	    if (rpmBase64Decode(key, (void **) &pkt, &pktlen) == 0) {
+	    if ((rc = rpmBase64Decode(key, (void **) &pkt, &pktlen)) == 0) {
 		rpmPubkey key = rpmPubkeyNew(pkt, pktlen);
 		int subkeysCount, i;
 		rpmPubkey *subkeys = rpmGetSubkeys(key, &subkeysCount);
 
-		if (rpmKeyringAddKey(ts->keyring, key) == 0) {
+		if ((rc = rpmKeyringAddKey(ts->keyring, key)) == 0) {
 		    rpmlog(RPMLOG_DEBUG, "added key %s to keyring\n", nvr);
 		    nkeys++;
 		}
@@ -357,12 +364,17 @@ static int loadKeyringFromDB(rpmts ts)
 			    i, nvr);
 
 			nkeys++;
+		    } else {
+			rc = 1;
 		    }
 		    rpmPubkeyFree(subkey);
 		}
 		free(subkeys);
 		free(pkt);
 	    }
+
+	    if (rc)
+		rpmlog(RPMLOG_WARNING, _("Could not load key %s\n"), nvr);
 	}
 	rpmtdFreeData(&pubkeys);
 	free(nvr);
-- 
2.51.1

